### ISSUE TITLE
CMake Build System

### ISSUE THREAD

**Issue by dwrj87** [moneymanagerex/moneymanagerex#679]
Would you be open to switching the build system to CMake?

On the plus side
- It generates both VS projects and Unix makefiles, amongst others
- Travis supports it so it could generate whatever it needs on-the-fly which means
- A lot of build artefacts could be taken out of the repo
- It would simplify the build instructions for Linux and make it a lot easier for me to maintain the PPAs.

On the downside
- Work involved, which I would be willing to do

**Comment by stef145g** [comment]
With reference to issue: https://github.com/moneymanagerex/moneymanagerex/issues/671, there is currently no ability for automatic builds for MSVC.2013 or MSVC.2015 as the current Bakefile 0.2.9 we are using no longer supports this.

**Bakefile 1.2.5** [https://github.com/vslavik/bakefile] requires rebuilding all the files.
Upon investigation it is missing features that does not allow for easy migration.

Another possibility was to convert to **premake5**. [https://github.com/premake/premake-core]
Again this would require a lot of work.

Since all upgrades require additional work, an upgrade to **CMake** would make sense especially because David (@dwrj87) is prepared to do this work.

I would sincerely recommend this move.

**Comment by siena123** [comment]
Pat has also mentioned the mac build is no longer correct from 0.2.9 bakefile being used. 

From the little bit of work that I have done on updating to the latest version of bakefile, I completely agree with you Stefano. Lets have David do the work to switch to CMake . 


- James

**Comment by dwrj87** [comment]
I have made some progress with this. I've got it compiling on Ubuntu with CMake, but there is still a lot to do so it may take a few weeks

I'll push the changes to https://github.com/dwrj87/moneymanagerex/tree/cmake and put in a pull request when it's more complete so that you guys can test it and pull it apart

**Comment by dwrj87** [comment]
Just an update - so far, on Linux, I've got it compiling, installing and creating a package, as well as compiling the tests. There are still some things I'm not happy about with the tests however so I'm still working on them

I should be ready to get it working on Windows within the week. I can't test OSX however, so I will do my best and hopefully somebody that can will take it over then to finish it off

**Comment by omalleypat** [comment]
> Pat has also mentioned the mac build is no longer correct from 0.2.9 bakefile being used.

Just a quick update - I did get bakefile 0.2.9 working on Mac over this past weekend.

**Comment by dwrj87** [comment]
I have now successfully compiled on Windows as well as Linux. I will do some tidying up over the next week or so and reopen a pull request

**Comment by vomikan** [comment]
I like the progress you've done.

**Comment by omalleypat** [comment]
@dwrj87 - can you provide instructions for using cmake? I'll fork your branch, but I don't know what to do to compile. Thanks.

**Comment by dwrj87** [comment]
On Linux and I presume Mac, I do. I've got it all set up to make debs and rpms for Linux, and it can do dmgs for Mac and nsis for Windows as well, but no point working on them until I know that it's building OK

- git clone https://github.com/dwrj87/moneymanagerex/tree/cmake
- cd moneymanagerex
- git checkout cmake
- git submodule update --init
- cmake .
- make

**Comment by omalleypat** [comment]
I need to be able to specify the path for my wxWidgets library. How would I do that? 
(cmake is telling me wxWidgets is not found, but I haven't told it where to look).

**Comment by omalleypat** [comment]
OK - I just modified the $PATH variable, seemed to work. I got up to:

```
[ 99%] Building CXX object CMakeFiles/mmex.dir/src/mac/platfdep.cpp.o
make[2]: *** No rule to make target `/Users/pom/Programming/wxWidgets/build-release/lib/libwx_osx_cocoau_scintilla-3.1.a', needed by `bin/mmex'.  Stop.
make[1]: *** [CMakeFiles/mmex.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by dwrj87** [comment]
Try now :) And what did you do to the $PATH?

**Comment by omalleypat** [comment]
Ok - I have a binary that opens now. I got the error that it can't find `bin/home_page.htt`. We will also need to get cmake to create an app bundle. Here was the structure with the old system (the binary is at mmex.app/Contents/MacOS/mmex - it looks like some of the stuff got duplicated though so this isn't 100% correct):

```
mmex.app/
└── Contents
    ├── Info.plist
    ├── MacOS
    │   └── mmex
    ├── Resources
    │   ├── ChartNew.js
    │   ├── home_page.htt
    │   ├── kaching.wav
    │   ├── master.css
    │   ├── mmdb.icns
    │   ├── mmex.icns
    │   ├── mmex.ico
    │   └── sorttable.js
    └── SharedSupport
        ├── applications
        │   └── mmex.desktop
        ├── contrib.txt
        ├── doc
        │   ├── contrib.txt
        │   ├── help
        │   │   ├── GRM.gif
        │   │   ├── backtotop.png
        │   │   ├── btn_donateCC_LG.gif
        │   │   ├── budget.html
        │   │   ├── budget_editor.png
        │   │   ├── budget_entry_details.png
        │   │   ├── budget_grid.png
        │   │   ├── contents.js
        │   │   ├── french
        │   │   │   ├── account.jpg
        │   │   │   ├── add_acc_wizard1.jpg
        │   │   │   ├── add_acc_wizard2.jpg
        │   │   │   ├── add_acc_wizard3.jpg
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   ├── main_screen.jpg
        │   │   │   └── mmex_concept.png
        │   │   ├── general_report_manager.html
        │   │   ├── hungarian
        │   │   │   ├── budget.html
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   └── mmex_concept_hu.png
        │   │   ├── index.html
        │   │   ├── italian
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   ├── mmex_concept.png
        │   │   │   ├── stock_dialog.jpg
        │   │   │   ├── stock_options.jpg
        │   │   │   └── stock_view.jpg
        │   │   ├── master.css
        │   │   ├── mmex_concept.png
        │   │   ├── mmex_db_structure.png
        │   │   ├── polish
        │   │   │   ├── GRM.gif
        │   │   │   ├── btn_donateCC_LG.gif
        │   │   │   ├── budget.html
        │   │   │   ├── budget_editor.png
        │   │   │   ├── budget_entry_details.png
        │   │   │   ├── budget_grid.png
        │   │   │   ├── general_report_manager.html
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   ├── mmex_concept.png
        │   │   │   ├── mmex_structure.jpg
        │   │   │   ├── stock_dialog.jpg
        │   │   │   ├── stock_options.jpg
        │   │   │   └── stock_view.jpg
        │   │   ├── russian
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   ├── mmex_concept.png
        │   │   │   └── stock_dialog.png
        │   │   ├── spanish
        │   │   │   ├── budget.html
        │   │   │   ├── index.html
        │   │   │   └── investment.html
        │   │   ├── stock_shares_new_shares.png
        │   │   ├── stock_shares_new_stock.png
        │   │   ├── stock_shares_portfolio.png
        │   │   ├── stock_shares_share_account.png
        │   │   ├── stock_shares_view_transactions.png
        │   │   └── stocks_and_shares.html
        │   ├── license.txt
        │   ├── mmex
        │   │   ├── contrib.txt
        │   │   ├── help
        │   │   │   ├── GRM.gif
        │   │   │   ├── backtotop.png
        │   │   │   ├── btn_donateCC_LG.gif
        │   │   │   ├── budget.html
        │   │   │   ├── budget_editor.png
        │   │   │   ├── budget_entry_details.png
        │   │   │   ├── budget_grid.png
        │   │   │   ├── french
        │   │   │   │   ├── account.jpg
        │   │   │   │   ├── add_acc_wizard1.jpg
        │   │   │   │   ├── add_acc_wizard2.jpg
        │   │   │   │   ├── add_acc_wizard3.jpg
        │   │   │   │   ├── index.html
        │   │   │   │   ├── investment.html
        │   │   │   │   ├── main_screen.jpg
        │   │   │   │   └── mmex_concept.png
        │   │   │   ├── general_report_manager.html
        │   │   │   ├── hungarian
        │   │   │   │   ├── budget.html
        │   │   │   │   ├── index.html
        │   │   │   │   ├── investment.html
        │   │   │   │   └── mmex_concept_hu.png
        │   │   │   ├── index.html
        │   │   │   ├── investment.html
        │   │   │   ├── italian
        │   │   │   │   ├── index.html
        │   │   │   │   ├── investment.html
        │   │   │   │   ├── mmex_concept.png
        │   │   │   │   ├── stock_dialog.jpg
        │   │   │   │   ├── stock_options.jpg
        │   │   │   │   └── stock_view.jpg
        │   │   │   ├── mmex_concept.png
        │   │   │   ├── mmex_db_structure.png
        │   │   │   ├── polish
        │   │   │   │   ├── GRM.gif
        │   │   │   │   ├── btn_donateCC_LG.gif
        │   │   │   │   ├── budget.html
        │   │   │   │   ├── budget_editor.png
        │   │   │   │   ├── budget_entry_details.png
        │   │   │   │   ├── budget_grid.png
        │   │   │   │   ├── general_report_manager.html
        │   │   │   │   ├── index.html
        │   │   │   │   ├── investment.html
        │   │   │   │   ├── mmex_concept.png
        │   │   │   │   ├── mmex_structure.jpg
        │   │   │   │   ├── stock_dialog.jpg
        │   │   │   │   ├── stock_options.jpg
        │   │   │   │   └── stock_view.jpg
        │   │   │   ├── russian
        │   │   │   │   ├── index.html
        │   │   │   │   ├── investment.html
        │   │   │   │   ├── mmex_concept.png
        │   │   │   │   └── stock_dialog.png
        │   │   │   ├── spanish
        │   │   │   │   ├── budget.html
        │   │   │   │   ├── index.html
        │   │   │   │   └── investment.html
        │   │   │   ├── stock_dialog.png
        │   │   │   ├── stock_options.png
        │   │   │   └── stock_view.png
        │   │   ├── license.txt
        │   │   └── version.txt
        │   └── version.txt
        ├── icons
        │   └── hicolor
        │       └── scalable
        │           └── apps
        │               └── mmex.svg
        ├── license.txt
        ├── man
        │   └── man1
        │       └── mmex.1.gz
        ├── po
        │   └── en
        │       ├── arabic.mo
        │       ├── bulgarian.mo
        │       ├── chinese_chs.mo
        │       ├── chinese_zh.mo
        │       ├── croatian.mo
        │       ├── czech.mo
        │       ├── dutch.mo
        │       ├── dutch_be.mo
        │       ├── english-uk.mo
        │       ├── english.mo
        │       ├── french.mo
        │       ├── german.mo
        │       ├── greek.mo
        │       ├── hebrew.mo
        │       ├── hindi.mo
        │       ├── hungarian.mo
        │       ├── indonesian.mo
        │       ├── italian.mo
        │       ├── japanese.mo
        │       ├── latvian.mo
        │       ├── lithuanian.mo
        │       ├── norwegian.mo
        │       ├── polish.mo
        │       ├── portuguese.mo
        │       ├── portuguese_br.mo
        │       ├── portuguese_portugal.mo
        │       ├── romanian.mo
        │       ├── russian.mo
        │       ├── serbian.mo
        │       ├── serbo-croatian.mo
        │       ├── slovak.mo
        │       ├── slovenian.mo
        │       ├── spanish.mo
        │       ├── swedish.mo
        │       ├── tamil.mo
        │       ├── turkish.mo
        │       ├── ukrainian.mo
        │       └── vietnamese.mo
        ├── res
        │   ├── Chart.js
        │   ├── ChartNew.js
        │   ├── home_page.htt
        │   ├── kaching.wav
        │   ├── master.css
        │   ├── mmex.ico
        │   └── sorttable.js
        └── version.txt

```

**Comment by dwrj87** [comment]
Yes, this only compiles, it doesn't put anything in the correct place. The make install and make package targets do this. Now I know it actually compiles on Mac I'll work on that next

**Comment by stef145g** [comment]
@dwrj87 I have installed CMake 3.6.0 which provides me with a user interface.
I have attempted to use CMake to generate the make files with the following results:
When I attempted to configure the build, I got:
<
-- Build Type = 
-- Building Tests = OFF
-- Compiler = MSVC
CMake Error at C:/Applications/cmake/share/cmake-3.6/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find Gettext (missing: GETTEXT_MSGMERGE_EXECUTABLE
  GETTEXT_MSGFMT_EXECUTABLE)
Call Stack (most recent call first):
  C:/Applications/cmake/share/cmake-3.6/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
  C:/Applications/cmake/share/cmake-3.6/Modules/FindGettext.cmake:87 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:41 (FIND_PACKAGE)
>

I then commented out line 41
Using the generator: 'Visual Studio 12 2013 Win64' I was able to generate the build files.

I then successfully loaded the generated files into VS 13 but could not produce a successful build.
